Mojang/DataFixerUpper - GroovyParsingException: Failed to parse build.gradle

## What version of OpenRewrite are you using?
I am using

- Gradle plugin 7.3.0
- Java Dependencies 1.31.0

## How are you running OpenRewrite?

Project (note the old revision): https://github.com/Mojang/DataFixerUpper/blob/c100ef03c2ab321e5de4f25ffbe277e924aa7ca5/build.gradle

refactorings.yml:
```
type: specs.openrewrite.org/v1beta/recipe
name: com.example.MyRecipe
recipeList:
- org.openrewrite.gradle.AddDependency: {groupId: org.mockito, artifactId: mockito-core,
    version: 5.16.0, onlyIfUsing: '*..*', configuration: testImplementation}
```

rewrite.gradle:
```
initscript {
  repositories {
    maven { url "https://plugins.gradle.org/m2" }
  }
  dependencies {
    classpath("org.openrewrite.rewrite:org.openrewrite.rewrite.gradle.plugin:7.3.0")
  }
}

rootProject {
  plugins.apply(org.openrewrite.gradle.RewritePlugin)

  repositories {
    maven { url "https://plugins.gradle.org/m2" }
  }

  dependencies {
    rewrite("org.openrewrite.recipe:rewrite-java-dependencies:1.31.0")
  }

  rewrite {
    activeRecipe("com.example.MyRecipe")
    activeStyle("org.openrewrite.java.Autodetect")
    configFile = System.getProperty("configFile")
  }

}
```

Command: `./gradlew :rewriteRun --init-script=rewrite.gradle -DconfigFile=refactorings.yml --info`


## What is the smallest, simplest way to reproduce the problem?

The above is as minimal as I've been able to get things, other than editing the original build.gradle.

## What did you expect to see?

The mockito dependency added to `build.gradle`

## What did you see instead?

An exception:
```
> Task :rewriteRun
Caching disabled for task ':rewriteRun' because:
  Build cache is disabled
Task ':rewriteRun' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
Validating active recipes
Scanning sources in project :
Using active styles []
Scanned 162 Java sources in :/main
Scanned 1 Java sources in :/test
Error during rewrite run
org.openrewrite.groovy.GroovyParsingException: Failed to parse build.gradle, cursor position likely inaccurate.
        at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:204)
        at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:147)
...
        at org.openrewrite.gradle.isolated.DefaultProjectParser.listResults(DefaultProjectParser.java:1382)
        at org.openrewrite.gradle.isolated.DefaultProjectParser.run(DefaultProjectParser.java:420)
        at org.openrewrite.gradle.DelegatingProjectParser.lambda$run$2(DelegatingProjectParser.java:115)
        at org.openrewrite.gradle.DelegatingProjectParser.unwrapInvocationException(DelegatingProjectParser.java:164)
        at org.openrewrite.gradle.DelegatingProjectParser.run(DelegatingProjectParser.java:114)
        at org.openrewrite.gradle.RewriteRunTask.run(RewriteRunTask.java:36)
...
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: -2
        at java.base/java.lang.StringLatin1.charAt(StringLatin1.java:48)
        at java.base/java.lang.String.charAt(String.java:1517)
        at org.openrewrite.groovy.GroovyParserVisitor.sourceSubstring(GroovyParserVisitor.java:2448)
        at org.openrewrite.groovy.GroovyParserVisitor.access$2500(GroovyParserVisitor.java:81)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.lambda$visitConstantExpression$3(GroovyParserVisitor.java:1261)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.insideParentheses(GroovyParserVisitor.java:660)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitConstantExpression(GroovyParserVisitor.java:1213)
        at org.codehaus.groovy.ast.expr.ConstantExpression.visit(ConstantExpression.java:80)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitArgumentlistExpression(GroovyParserVisitor.java:740)
        at org.codehaus.groovy.ast.expr.ArgumentListExpression.visit(ArgumentListExpression.java:73)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.lambda$visitMethodCallExpression$10(GroovyParserVisitor.java:1681)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.insideParentheses(GroovyParserVisitor.java:653)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitMethodCallExpression(GroovyParserVisitor.java:1606)
        at org.codehaus.groovy.ast.expr.MethodCallExpression.visit(MethodCallExpression.java:76)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitDeclarationExpression(GroovyParserVisitor.java:1380)
        at org.codehaus.groovy.ast.expr.DeclarationExpression.visit(DeclarationExpression.java:89)
        at org.codehaus.groovy.ast.CodeVisitorSupport.visitExpressionStatement(CodeVisitorSupport.java:117)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.lambda$visitExpressionStatement$6(GroovyParserVisitor.java:1421)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.labeled(GroovyParserVisitor.java:675)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitExpressionStatement(GroovyParserVisitor.java:1420)
        at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:40)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:989)
        at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:69)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitIfElse(GroovyParserVisitor.java:1501)
        at org.codehaus.groovy.ast.stmt.IfStatement.visit(IfStatement.java:41)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:989)
        at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:69)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitClosureExpression(GroovyParserVisitor.java:1183)
        at org.codehaus.groovy.ast.expr.ClosureExpression.visit(ClosureExpression.java:46)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitArgumentlistExpression(GroovyParserVisitor.java:740)
        at org.codehaus.groovy.ast.expr.ArgumentListExpression.visit(ArgumentListExpression.java:73)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.lambda$visitMethodCallExpression$10(GroovyParserVisitor.java:1681)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.insideParentheses(GroovyParserVisitor.java:653)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitMethodCallExpression(GroovyParserVisitor.java:1606)
        at org.codehaus.groovy.ast.expr.MethodCallExpression.visit(MethodCallExpression.java:76)
        at org.codehaus.groovy.ast.CodeVisitorSupport.visitExpressionStatement(CodeVisitorSupport.java:117)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.lambda$visitExpressionStatement$6(GroovyParserVisitor.java:1421)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.labeled(GroovyParserVisitor.java:675)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitExpressionStatement(GroovyParserVisitor.java:1420)
        at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:40)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:989)
        at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:69)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitClosureExpression(GroovyParserVisitor.java:1183)
        at org.codehaus.groovy.ast.expr.ClosureExpression.visit(ClosureExpression.java:46)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitArgumentlistExpression(GroovyParserVisitor.java:740)
        at org.codehaus.groovy.ast.expr.ArgumentListExpression.visit(ArgumentListExpression.java:73)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.lambda$visitMethodCallExpression$10(GroovyParserVisitor.java:1681)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.insideParentheses(GroovyParserVisitor.java:653)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitMethodCallExpression(GroovyParserVisitor.java:1606)
        at org.codehaus.groovy.ast.expr.MethodCallExpression.visit(MethodCallExpression.java:76)
        at org.codehaus.groovy.ast.CodeVisitorSupport.visitExpressionStatement(CodeVisitorSupport.java:117)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.lambda$visitExpressionStatement$6(GroovyParserVisitor.java:1421)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.labeled(GroovyParserVisitor.java:675)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitExpressionStatement(GroovyParserVisitor.java:1420)
        at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:40)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:989)
        at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:69)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:619)
        at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitIfElse(GroovyParserVisitor.java:1504)
        at org.codehaus.groovy.ast.stmt.IfStatement.visit(IfStatement.java:41)
        at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:2148)
        at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:196)
        ... 170 more
There were problems parsing some source files, run with --info to see full stack traces
There were problems parsing build.gradle
```

## What is the full stack trace of any errors you encountered?

See above for stack trace with java.base and org.gradle elements dropped.

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?

Don't have active plans to contribute a fix but if there's some guidance available then am happy to consider it.

--- a/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
+++ b/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
@@ -2442,6 +2442,25 @@ public class GroovyParserVisitor {
         return false;
     }
 
+
+    /**
+     * Determines if the character at the specified {@code index} in the {@code source} is escaped.
+     * A character is considered escaped if it is preceded by an odd number of consecutive backslashes.
+     *
+     * @param index The index in the string to check for escaping.
+     *              This must be a valid index within the range of {@code source}.
+     * @return {@code true} if the character at the given index is escaped, {@code false} otherwise.
+     * @throws StringIndexOutOfBoundsException If the {@code index} is not within the valid range of the source string.
+     */
+    private boolean isEscaped(int index) {
+        int backslashCount = 0;
+        while (index >= 0 && source.charAt(index) == '\\') {
+            backslashCount++;
+            index--;
+        }
+        return backslashCount % 2 != 0;
+    }
+
     /**
      * Returns a string that is a part of this source. The substring begins at the specified beginIndex and extends until delimiter.
      * The cursor will not be moved.
@@ -2449,7 +2468,8 @@ public class GroovyParserVisitor {
     private String sourceSubstring(int beginIndex, String untilDelim) {
         int endIndex = source.indexOf(untilDelim, Math.max(beginIndex, cursor + untilDelim.length()));
         // don't stop if last char is escaped.
-        while (source.charAt(endIndex - 1) == '\\') {
+        // Fixed potential infinite loop by correctly handling escaped delimiters in the source string.
+        while (endIndex > 0 && isEscaped(endIndex - 1)) {
             endIndex = source.indexOf(untilDelim, endIndex + 1);
         }
         return source.substring(beginIndex, endIndex);

--- a/rewrite-groovy/src/test/java/org/openrewrite/groovy/tree/RealWorldGroovyTest.java
+++ b/rewrite-groovy/src/test/java/org/openrewrite/groovy/tree/RealWorldGroovyTest.java
@@ -441,4 +441,16 @@ class RealWorldGroovyTest implements RewriteTest {
           )
         );
     }
+
+    @Test
+    void stringWithMultipleBackslashes() {
+        rewriteRun(
+          groovy(
+            """
+              "".replaceAll('\\\\', '/')
+              "a\\b".replaceAll('\\\\', '/')
+              """
+          )
+        );
+    }
 }

